x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SLMGetExpiredSnapshotsAction.java

```diff
@@ -0,0 +1,312 @@
+            final var resultsBuilder = new ResultsBuilder();
+            try (var refs = new RefCountingRunnable(() -> listener.onResponse(resultsBuilder.getResponse()))) {
+                for (final var repositoryName : request.repositories()) {
+
+                    final Repository repository;
+                    try {
+                        repository = repositoriesService.repository(repositoryName);
+                    } catch (RepositoryMissingException e) {
+                        logger.debug("[{}]: repository not found", repositoryName);
+                        continue;
+                    }
+
+                    if (repository.isReadOnly()) {
+                        logger.debug("[{}]: skipping readonly repository", repositoryName);
+                        continue;
+                    }
+
+                    retentionExecutor.execute(
+                        ActionRunnable.wrap(
+                            refs.acquireListener(),
+                            perRepositoryListener -> SubscribableListener
+
+                                // Get repository data
+                                .newForked(repository::getRepositoryData)
+
+                                // Collect snapshot details by policy, and get any missing details by reading SnapshotInfo
+                                .<SnapshotDetailsByPolicy>andThen(
+                                    retentionExecutor,
+                                    threadContext,
+                                    (l, repositoryData) -> getSnapshotDetailsByPolicy(
+                                        repositoriesService.repository(repositoryName),
+                                        repositoryData,
+                                        l
+                                    )
+                                )
+
+                                // Compute snapshots to delete for each (relevant) policy
+                                .<Void>andThen(
+                                    retentionExecutor,
+                                    threadContext,
+                                    (l, snapshotDetailsByPolicy) -> ActionListener.completeWith(l, () -> {
+                                        resultsBuilder.addResult(
+                                            repositoryName,
+                                            getSnapshotsToDelete(repositoryName, request.policies(), snapshotDetailsByPolicy)
+                                        );
+                                        return null;
+                                    })
+                                )
+
+                                // And notify this repository's listener on completion
+                                .addListener(perRepositoryListener.delegateResponse((l, e) -> {
+                                    logger.debug(Strings.format("[%s]: could not compute expired snapshots", repositoryName), e);
```

Comment:
Almost certainly my own misunderstanding, but why is it that this `logger.debug` isn't just always called? To my unfamiliar eyes, it looks like we'd always just log the message for all repositories.

Reply:
It's under `delegateResponse` so it's only called if the listener is completed exceptionally. FTR I do not like the name but it is very well-established.

Reply:
Ah yup, I got it now, thanks!